samza-core/src/main/scala/org/apache/samza/storage/TaskStorageManager.scala

```diff
@@ -228,11 +230,28 @@ class TaskStorageManager(
 
     for ((storeName, systemStream) <- changeLogSystemStreams) {
       val systemStreamPartition = new SystemStreamPartition(systemStream, partition)
+      val admin = systemAdmins.getOrElse(systemStream.getSystem,
+        throw new SamzaException("Unable to get systemAdmin for store " + storeName + " and systemStream" + systemStream))
       val consumer = storeConsumers(storeName)
-      val offset =
-        Option(fileOffset.get(systemStreamPartition))
-          .getOrElse(changeLogOldestOffsets
-            .getOrElse(systemStream, throw new SamzaException("Missing a change log offset for %s." format systemStreamPartition)))
+      val fileOffset = fileOffsets.get(systemStreamPartition)
+      val oldestOffset = changeLogOldestOffsets
+        .getOrElse(systemStream, throw new SamzaException("Missing a change log offset for %s." format systemStreamPartition));
+      val offset = if (fileOffset != null) {
+        // File offset was the last message written to the changelog that is also reflected in the store,
+        // so we start with the NEXT offset
+        val resumeOffset = admin.getOffsetsAfter(Map(systemStreamPartition -> fileOffset).asJava).get(systemStreamPartition)
```

Comment:
Minor: IMHO it's worth extracting this as a private helper method (e.g. getStartingOffset(oldestOffset, fileOffset)) with documentation + tests.

Reply:
Totally agree. I've actually already done this on my local copy. Will update the PR when the unit tests are done.